File: common/src/main/java/de/hype/bingonet/environment/addonpacketconfig/AddonPacketManager.java
package de.hype.bingonet.environment.addonpacketconfig;

import de.hype.bingonet.client.common.client.socketAddons.AddonHandler;
import de.hype.bingonet.shared.packets.addonpacket.*;

import java.util.ArrayList;
import java.util.List;

public class AddonPacketManager {

    private static AddonPacketManager lastAddonPacketManager = null;
    List<AddonPacket<? extends AbstractAddonPacket>> addonPackets = new ArrayList<>();

    public List<AddonPacket<? extends AbstractAddonPacket>> getPackets() {
        return addonPackets;
    }

    // Define a map to store packet classes and their associated actions
    AddonHandler connection;

    // Method to initialize packet actions
    public AddonPacketManager(AddonHandler connection) {
        this.connection = connection;
        initializePacketActions(connection);
        lastAddonPacketManager = this;
    }

    public void initializePacketActions(AddonHandler connection) {
        addonPackets.add(new AddonPacket<>(DisplayTellrawMessageAddonPacket.class, connection::onDisplayTellrawMessageAddonPacket));
        addonPackets.add(new AddonPacket<>(DisplayClientsideMessageAddonPacket.class, connection::onDisplayClientsideMessageAddonPacket));
        addonPackets.add(new AddonPacket<>(PlaySoundAddonPacket.class, connection::onPlaySoundAddonPacket));
        addonPackets.add(new AddonPacket<>(PublicChatAddonPacket.class, connection::onPublicChatAddonPacket));
        addonPackets.add(new AddonPacket<>(ServerCommandAddonPacket.class, connection::onServerCommandAddonPacket));
        addonPackets.add(new AddonPacket<>(ChatPromptAddonPacket.class, connection::onChatPromptAddonPacket));
        addonPackets.add(new AddonPacket<>(WaypointAddonPacket.class, connection::onWaypointAddonPacket));
        addonPackets.add(new AddonPacket<>(GetWaypointsAddonPacket.class, connection::onGetWaypointsAddonPacket));
        addonPackets.add(new AddonPacket<>(ClientCommandAddonPacket.class, connection::onClientCommandAddonPacket));
//        addonPackets.add(new AddonPacket<>(ReceivedPublicChatMessageAddonPacket.class, connection::onReceivedPublicChatMessageAddonPacket));
        addonPackets.add(new AddonPacket<>(StatusUpdateAddonPacket.class, connection::onStatusUpdateAddonPacket));


    }

    // Method to handle a received packet


    //   method to get a list of all packets
    public static List<Class<? extends AbstractAddonPacket>> getAllPacketClasses() {
        if (lastAddonPacketManager == null) {
            lastAddonPacketManager = new AddonPacketManager(null);
        }
        List<Class<? extends AbstractAddonPacket>> allPackets = new ArrayList<>();
        for (int i = 0; i < lastAddonPacketManager.addonPackets.size(); i++) {
            allPackets.add(lastAddonPacketManager.addonPackets.get(i).getClazz());
        }
        return allPackets;
    }
}

File: common/src/main/java/de/hype/bingonet/environment/addonpacketconfig/AddonPacketManager.kt
package de.hype.bingonet.environment.addonpacketconfig

import de.hype.bingonet.client.common.client.socketAddons.AddonHandler
import de.hype.bingonet.shared.packets.addonpacket.*

class AddonPacketManager(// Define a map to store packet classes and their associated actions
    var connection: AddonHandler?
) {
    var packets: MutableList<AddonPacket<out AbstractAddonPacket>> =
        ArrayList()

    // Method to initialize packet actions
    init {
        initializePacketActions(connection)
        lastAddonPacketManager = this
    }

    fun initializePacketActions(connection: AddonHandler?) {
        packets.add(
            AddonPacket(
                DisplayTellrawMessageAddonPacket::class.java
            ) { packet: DisplayTellrawMessageAddonPacket ->
                connection?.onDisplayTellrawMessageAddonPacket(packet)
            }
        )
        packets.add(
            AddonPacket(
                DisplayClientsideMessageAddonPacket::class.java
            ) { packet: DisplayClientsideMessageAddonPacket ->
                connection?.onDisplayClientsideMessageAddonPacket(packet)
            }
        )
        packets.add(
            AddonPacket(
                PlaySoundAddonPacket::class.java
            ) { packet: PlaySoundAddonPacket ->
                connection?.onPlaySoundAddonPacket(packet)
            }
        )
        packets.add(
            AddonPacket(
                PublicChatAddonPacket::class.java
            ) { packet: PublicChatAddonPacket ->
                connection?.onPublicChatAddonPacket(packet)
            }
        )
        packets.add(
            AddonPacket(
                ServerCommandAddonPacket::class.java
            ) { packet: ServerCommandAddonPacket ->
                connection?.onServerCommandAddonPacket(packet)
            }
        )
        packets.add(
            AddonPacket(
                ChatPromptAddonPacket::class.java
            ) { packet: ChatPromptAddonPacket ->
                connection?.onChatPromptAddonPacket(packet)
            }
        )
        packets.add(
            AddonPacket(
                WaypointAddonPacket::class.java
            ) { packet: WaypointAddonPacket ->
                connection?.onWaypointAddonPacket(packet)
            }
        )
        packets.add(
            AddonPacket(
                GetWaypointsAddonPacket::class.java
            ) { packet: GetWaypointsAddonPacket -> connection?.onGetWaypointsAddonPacket(packet) }
        )
        packets.add(
            AddonPacket(
                ClientCommandAddonPacket::class.java
            ) { packet: ClientCommandAddonPacket ->
                connection?.onClientCommandAddonPacket(packet)
            }
        )
        packets.add(
            AddonPacket(
                GoToIslandAddonPacket::class.java
            ) { packet: GoToIslandAddonPacket ->
                connection?.onGoToIslandAddonPacket(packet)
            }
        )
        //        addonPackets.add(new AddonPacket<>(ReceivedPublicChatMessageAddonPacket.class, connection::onReceivedPublicChatMessageAddonPacket));
        packets.add(
            AddonPacket(
                StatusUpdateAddonPacket::class.java
            ) { packet: StatusUpdateAddonPacket -> connection?.onStatusUpdateAddonPacket(packet) }
        )
    }


    companion object {
        private var lastAddonPacketManager: AddonPacketManager? = null
        val allPacketClasses: MutableList<Class<out AbstractAddonPacket>>
            // Method to handle a received packet
            get() {
                if (lastAddonPacketManager == null) {
                    lastAddonPacketManager = AddonPacketManager(null)
                }
                val allPackets: MutableList<Class<out AbstractAddonPacket>> =
                    ArrayList()
                for (i in lastAddonPacketManager!!.packets.indices) {
                    allPackets.add(lastAddonPacketManager!!.packets.get(i).clazz)
                }
                return allPackets
            }
    }
}
